File: algorithms/code/leetcode/lc275_h_index_ii/lc275_h_index_ii.py
# https://leetcode.com/problems/h-index-ii/

import unittest


class Solution:
    def hIndex(self, citations: list[int]) -> int:
        n = len(citations)
        a = 0
        z = n
        while a < z:
            m = (a + z) >> 1
            if citations[m] == n - m:
                return n - m
            if citations[m] < n - m:
                a = m + 1
            else:
                z = m
        return n - a


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        citations = [0, 1, 3, 5, 6]
        self.assertEqual(3, Solution().hIndex(citations))

    def test_none(self) -> None:
        citations = [0, 0, 0, 0, 0]
        self.assertEqual(0, Solution().hIndex(citations))

    def test_100(self) -> None:
        citations = [100]
        self.assertEqual(1, Solution().hIndex(citations))

File: algorithms/code/leetcode/lc275_h_index_ii/lc275_h_index_ii.rs
// https://leetcode.com/problems/h-index-ii/
// #medium

use std::cmp::Ordering;

pub struct Solution;

impl Solution {
    pub fn h_index(citations: Vec<i32>) -> i32 {
        let n = citations.len() as i32;
        let mut a = 0;
        let mut z = n;
        while a < z {
            let m = (a + z) / 2;
            match citations[m as usize].cmp(&(n - m)) {
                Ordering::Equal => {
                    return n - m;
                }
                Ordering::Less => {
                    a = m + 1;
                }
                Ordering::Greater => {
                    z = m;
                }
            }
        }
        n - a
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        assert_eq!(Solution::h_index(vec![0, 1, 3, 5, 6]), 3);
    }

    #[test]
    fn test_none() {
        assert_eq!(Solution::h_index(vec![0, 0, 0, 0, 0]), 0);
    }

    #[test]
    fn test_100() {
        assert_eq!(Solution::h_index(vec![100]), 1);
    }
}
